File: web/src/collision/move.ts
// Boxland — canonical swept-AABB collision (web port).
//
// THIS IMPLEMENTATION IS A LITERAL PORT of schemas/collision.md. Server
// (Go), web (this), and (at v1.1) iOS (Swift) MUST produce byte-identical
// resolved deltas for every vector in /shared/test-vectors/collision.json.
// CI gates all runtimes against the same corpus.
//
// All math is integer arithmetic on int32 fixed-point sub-pixels (1 px =
// 256 sub-units). No floats anywhere on the hot path.

import {
	EDGE_E,
	EDGE_N,
	EDGE_S,
	EDGE_W,
	TILE_SIZE_SUB,
	isOneWay,
	type AABB,
	type Entity,
	type MoveResult,
	type World,
} from "./types";

// Axis is a tagged number so signature reads are obvious.
const AXIS_X = 0;
const AXIS_Y = 1;
type Axis = typeof AXIS_X | typeof AXIS_Y;

/**
 * Move `entity` by (dx, dy) sub-pixels through `world`, applying axis-
 * separated swept-AABB collision. The entity's aabb is mutated in place to
 * reflect the resolved position. Returns the actual delta applied so
 * callers can observe slides.
 */
export function move(entity: Entity, dx: number, dy: number, world: World): MoveResult {
	const dxResolved = sweepAxis(entity, AXIS_X, dx | 0, world);
	const dyResolved = sweepAxis(entity, AXIS_Y, dy | 0, world);
	return { resolvedDx: dxResolved, resolvedDy: dyResolved };
}

function sweepAxis(entity: Entity, axis: Axis, step: number, world: World): number {
	if (step === 0) return 0;

	const sweepBox = extendAABB(entity.aabb, axis, step);
	const [gx0, gy0, gx1, gy1] = tileRangeOverlapping(sweepBox);

	const sign = step > 0 ? 1 : -1;
	const edgeBit = facingEdgeBit(axis, sign);

	let blockedAt = step;

	for (let gy = gy0; gy <= gy1; gy++) {
		for (let gx = gx0; gx <= gx1; gx++) {
			const T = world.get(gx, gy);
			if (!T) continue;
			if ((T.collision_layer_mask & entity.mask) === 0) continue;
			if ((T.edge_collisions & edgeBit) === 0) continue;

			// One-way platform rule: a tile authored as OneWayN blocks
			// downward motion ONLY when the entity's foot is already at
			// or above the tile top. Otherwise pass through (entry from
			// the side, or rising from below). Mirror of the Go rule
			// in server/internal/sim/collision/move.go.
			if (isOneWay(T.shape)) {
				if (axis !== AXIS_Y || sign <= 0) continue;
				const tileTop = gy * TILE_SIZE_SUB;
				if (entity.aabb.bottom > tileTop) continue;
			}

			const contact = distanceToEdge(entity.aabb, gx, gy, axis, sign);
			if (sign > 0) {
				const clamped = contact < 0 ? 0 : contact;
				if (clamped < blockedAt) blockedAt = clamped;
			} else {
				const clamped = contact > 0 ? 0 : contact;
				if (clamped > blockedAt) blockedAt = clamped;
			}
		}
	}

	advance(entity.aabb, axis, blockedAt);
	return blockedAt;
}

// ---- helpers ----

function extendAABB(box: AABB, axis: Axis, step: number): AABB {
	if (axis === AXIS_X) {
		return step >= 0
			? { left: box.left, top: box.top, right: box.right + step, bottom: box.bottom }
			: { left: box.left + step, top: box.top, right: box.right, bottom: box.bottom };
	}
	return step >= 0
		? { left: box.left, top: box.top, right: box.right, bottom: box.bottom + step }
		: { left: box.left, top: box.top + step, right: box.right, bottom: box.bottom };
}

/**
 * Tile-grid AABB cover. Uses inclusive bounds; the right/bottom edge of a
 * box that sits exactly on a tile boundary belongs to the *previous* tile,
 * not the next, to avoid touching a tile we don't actually overlap.
 */
function tileRangeOverlapping(box: AABB): [number, number, number, number] {
	const gx0 = Math.floor(box.left   / TILE_SIZE_SUB);
	const gy0 = Math.floor(box.top    / TILE_SIZE_SUB);
	const gx1 = Math.floor((box.right  - 1) / TILE_SIZE_SUB);
	const gy1 = Math.floor((box.bottom - 1) / TILE_SIZE_SUB);
	return [gx0, gy0, gx1, gy1];
}

function facingEdgeBit(axis: Axis, sign: number): number {
	if (axis === AXIS_X) return sign > 0 ? EDGE_W : EDGE_E;
	return sign > 0 ? EDGE_N : EDGE_S;
}

function distanceToEdge(aabb: AABB, gx: number, gy: number, axis: Axis, sign: number): number {
	const tLeft   = gx * TILE_SIZE_SUB;
	const tTop    = gy * TILE_SIZE_SUB;
	const tRight  = tLeft + TILE_SIZE_SUB;
	const tBottom = tTop + TILE_SIZE_SUB;
	if (axis === AXIS_X) {
		return sign > 0 ? tLeft - aabb.right : tRight - aabb.left;
	}
	return sign > 0 ? tTop - aabb.bottom : tBottom - aabb.top;
}

function advance(aabb: AABB, axis: Axis, by: number): void {
	if (by === 0) return;
	if (axis === AXIS_X) {
		aabb.left += by;
		aabb.right += by;
	} else {
		aabb.top += by;
		aabb.bottom += by;
	}
}

File: server/internal/sim/collision/move.go
// Boxland — canonical swept-AABB collision (server port).
//
// THIS IMPLEMENTATION IS A LITERAL PORT of schemas/collision.md and must
// produce byte-identical resolved deltas to web/src/collision/move.ts for
// every vector in /shared/test-vectors/collision.json. The corpus is
// gated against this implementation in collision_corpus_test.go.
//
// Algorithm (axis-separated swept AABB):
//
//	move(entity, Δ):
//	  for axis in [X, Y]:
//	    sweep entity.AABB along Δ[axis]
//	    for each tile T overlapping the sweep:
//	      if (T.collision_layer_mask & entity.mask) == 0: continue
//	      determine which edge of T faces the sweep direction
//	      if that edge bit is unset: continue
//	      clip Δ[axis] to the contact point
//	    apply (clipped) Δ[axis]
package collision

const (
	axisX = 0
	axisY = 1
)

// Move advances entity by (dx, dy) sub-pixels, applying axis-separated
// swept-AABB collision against world. Mutates entity.AABB in place;
// returns the actual delta applied so callers can observe slides.
func Move(entity *Entity, dx, dy int32, world World) MoveResult {
	dxR := sweepAxis(entity, axisX, dx, world)
	dyR := sweepAxis(entity, axisY, dy, world)
	return MoveResult{ResolvedDX: dxR, ResolvedDY: dyR}
}

func sweepAxis(entity *Entity, axis int, step int32, world World) int32 {
	if step == 0 {
		return 0
	}

	sweep := extendAABB(entity.AABB, axis, step)
	gx0, gy0, gx1, gy1 := tileRangeOverlapping(sweep)

	sign := int32(1)
	if step < 0 {
		sign = -1
	}
	edgeBit := facingEdgeBit(axis, sign)

	blockedAt := step

	for gy := gy0; gy <= gy1; gy++ {
		for gx := gx0; gx <= gx1; gx++ {
			t, ok := world.TileAt(gx, gy)
			if !ok {
				continue
			}
			if (t.CollisionLayerMask & entity.Mask) == 0 {
				continue
			}
			if (t.EdgeCollisions & edgeBit) == 0 {
				continue
			}
			// One-way platform rule: a tile authored as ShapeOneWayN
			// blocks downward motion (axis=Y, sign>0) ONLY when the
			// entity's foot is already at or above the tile top at the
			// START of the sweep. Otherwise skip the block — entities
			// crossing in from the side or rising through the bottom
			// pass through cleanly. See shape.go IsOneWay.
			if IsOneWay(t.Shape) {
				if axis != axisY || sign <= 0 {
					continue
				}
				tileTop := gy * TileSizeSub
				if entity.AABB.Bottom > tileTop {
					continue
				}
			}
			contact := distanceToEdge(entity.AABB, gx, gy, axis, sign)
			if sign > 0 {
				clamped := contact
				if clamped < 0 {
					clamped = 0
				}
				if clamped < blockedAt {
					blockedAt = clamped
				}
			} else {
				clamped := contact
				if clamped > 0 {
					clamped = 0
				}
				if clamped > blockedAt {
					blockedAt = clamped
				}
			}
		}
	}

	advance(&entity.AABB, axis, blockedAt)
	return blockedAt
}

// ---- helpers ----

func extendAABB(box AABB, axis int, step int32) AABB {
	if axis == axisX {
		if step >= 0 {
			return AABB{Left: box.Left, Top: box.Top, Right: box.Right + step, Bottom: box.Bottom}
		}
		return AABB{Left: box.Left + step, Top: box.Top, Right: box.Right, Bottom: box.Bottom}
	}
	if step >= 0 {
		return AABB{Left: box.Left, Top: box.Top, Right: box.Right, Bottom: box.Bottom + step}
	}
	return AABB{Left: box.Left, Top: box.Top + step, Right: box.Right, Bottom: box.Bottom}
}

// tileRangeOverlapping returns the (inclusive) tile-grid bounds the AABB
// touches. Right/bottom edges that lie exactly on a tile boundary belong
// to the previous tile, not the next, to avoid touching a tile we don't
// actually overlap.
func tileRangeOverlapping(box AABB) (gx0, gy0, gx1, gy1 int32) {
	gx0 = floorDiv(box.Left, TileSizeSub)
	gy0 = floorDiv(box.Top, TileSizeSub)
	gx1 = floorDiv(box.Right-1, TileSizeSub)
	gy1 = floorDiv(box.Bottom-1, TileSizeSub)
	return
}

func facingEdgeBit(axis int, sign int32) uint8 {
	if axis == axisX {
		if sign > 0 {
			return EdgeW
		}
		return EdgeE
	}
	if sign > 0 {
		return EdgeN
	}
	return EdgeS
}

func distanceToEdge(aabb AABB, gx, gy int32, axis int, sign int32) int32 {
	tLeft := gx * TileSizeSub
	tTop := gy * TileSizeSub
	tRight := tLeft + TileSizeSub
	tBottom := tTop + TileSizeSub
	if axis == axisX {
		if sign > 0 {
			return tLeft - aabb.Right
		}
		return tRight - aabb.Left
	}
	if sign > 0 {
		return tTop - aabb.Bottom
	}
	return tBottom - aabb.Top
}

func advance(aabb *AABB, axis int, by int32) {
	if by == 0 {
		return
	}
	if axis == axisX {
		aabb.Left += by
		aabb.Right += by
		return
	}
	aabb.Top += by
	aabb.Bottom += by
}

// floorDiv is integer division rounded towards negative infinity.
// Stdlib `/` rounds towards zero; we need floor for chunk math.
func floorDiv(a, b int32) int32 {
	q := a / b
	if (a%b != 0) && ((a < 0) != (b < 0)) {
		q--
	}
	return q
}
